### PR TITLE
[SPARK-18342] Make rename failures fatal in HDFSBackedStateStore

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

If the rename operation in the state store fails (`fs.rename` returns `false`), the StateStore should throw an exception and have the task retry. Currently if renames fail, nothing happens during execution immediately. However, you will observe that snapshot operations will fail, and then any attempt at recovery (executor failure / checkpoint recovery) also fails.

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @tdas

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 1]
```diff
@@ -525,7 +527,7 @@ private[state] class HDFSBackedStateStoreProvider(
 
         val deltaFiles = allFiles.filter { file =>
           file.version > snapshotFile.version && file.version <= version
-        }
+        }.toList
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: the line returns a `Stream` which is a Scala Iterator like class, therefore the logging below doesn't work

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala [hunk 2]
```diff
@@ -598,3 +612,19 @@ private[state] object StateStoreSuite {
     }}.toSet
   }
 }
+
+/** Fake FileSystem to test whether the method `fs.exists` is called during
+ * `DataSource.resolveRelation`.
+ */
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala]: Wrong docs.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala [hunk 3]
```diff
@@ -455,6 +456,18 @@ class StateStoreSuite extends SparkFunSuite with BeforeAndAfter with PrivateMeth
     }
   }
 
+  test("commit fails when rename fails") {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StateStoreSuite.scala]: can you add the jira number here.

**Comment by tdas** [comment]
looks good. to minor comments.

**Comment by brkyvz** [comment]
@tdas Addressed

**Comment by tdas** [comment]
LGTM.

**Comment by tdas** [comment]
Merge this to master and 2.0

**Pull request closed without merging**